src/main/java/org/apache/commons/text/numbers/ParsedDecimal.java

```diff
@@ -0,0 +1,715 @@
+        int nonZeroCount = count;
+        for (int i = count - 1;
+                i >= 0 && digits[i] == 0;
+                --i) {
+            --nonZeroCount;
+        }
+        exponent += digitCount - nonZeroCount;
+        digitCount = nonZeroCount;
+    }
+
+    /** Construct a new instance from the given double value.
+     * @param d double value
+     * @return a new instance containing the parsed components of the given double value
+     * @throws IllegalArgumentException if {@code d} is {@code NaN} or infinite
+     */
+    public static ParsedDecimal from(final double d) {
+        if (!Double.isFinite(d)) {
+            throw new IllegalArgumentException("Double is not finite");
+        }
+
+        // Get the canonical string representation of the double value and parse
+        // it to extract the components of the decimal value. From the documentation
+        // of Double.toString() and the fact that d is finite, we are guaranteed the
+        // following:
+        // - the string will not be empty
+        // - it will contain exactly one decimal point character
+        // - all digit characters are in the ASCII range
+        final char[] strChars = Double.toString(d).toCharArray();
+
+        final boolean negative = strChars[0] == MINUS_CHAR;
+        final int digitStartIdx = negative ? 1 : 0;
+
+        final int[] digits = new int[strChars.length];
+
+        boolean foundDecimalPoint = false;
+        int digitCount = 0;
+        int significantDigitCount = 0;
+        int decimalPos = 0;
+
+        int i;
+        for (i = digitStartIdx; i < strChars.length; ++i) {
+            final char ch = strChars[i];
+
+            if (ch == DECIMAL_SEP_CHAR) {
+                foundDecimalPoint = true;
+                decimalPos = digitCount;
+            } else if (ch == EXPONENT_CHAR) {
+                // no more mantissa digits
+                break;
+            } else if (ch != ZERO_CHAR || digitCount > 0) {
+                // this is either the first non-zero digit or one after it
+                final int val = digitValue(ch);
+                digits[digitCount++] = val;
+
+                if (val > 0) {
+                    significantDigitCount = digitCount;
+                }
+            } else if (foundDecimalPoint) {
+                // leading zero in a fraction; adjust the decimal position
+                --decimalPos;
+            }
+        }
+
+        if (digitCount > 0) {
+            // determine the exponent
+            final int explicitExponent = i < strChars.length
+                    ? parseExponent(strChars, i + 1)
+                    : 0;
+            final int exponent = explicitExponent + decimalPos - significantDigitCount;
+
+            return new ParsedDecimal(negative, digits, significantDigitCount, exponent);
+        }
+
+        // no non-zero digits, so value is zero
+        return new ParsedDecimal(negative, new int[] {0}, 1, 0);
```

Comment:
A case for a singleton specialisation here.

Reply:
I made `ParsedDecimal` instances mutable so that rounding operations can modify the digit array in place instead of creating a copy. So, I don't see how a singleton would work here.

Reply:
Is this not zero (i.e. no non-zero digits)? Does the output differ when formatting parameters are changed for the value zero beyond putting extra zeros on:
```
0.0
0.00
0.000
0.0000e0
0.00000e000
```
Is any rounding possible when the value is zero?

You could have a singleton for negative and positive zero as this is something supported by a double.

Reply:
All of the examples you gave there will be interpreted as zero. Also, no additional rounding is possible when the instance is zero. Regardless, it still feels weird to have singletons of mutable instances. Is this what you're proposing here?

Reply:
The examples were of output that could be possible with a zero depending on formatting options. They were not meant to be input. The intension was to ask if the formatting supports this output for zero. Then a singleton for zero would have to respond appropriately.

`"it still feels weird to have singletons of mutable instances"`

It is not any weirder than having a mutable instance that cannot have its state changed. Which is my point. If you can never change the state of the `ParsedDecimal` for zero then surely this is the case for a singleton. Not having read the entire code I do not know if this is true. So can you change the state of a zero?

Reply:
> The examples were of output that could be possible with a zero depending on formatting options.

The only possible output formats for zero are listed below. Each form could also be prefixed with a minus sign if needed and allowed by the configuration.
```
0
0.0
0E0
0.0E0
```
> So can you change the state of a zero?

Not the numeric state. However, each instance has an internal char buffer for building output strings. So, the instances are not thread-safe.